staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -535,10 +535,12 @@ func (w *watchCache) list(ctx context.Context, resourceVersion uint64, key strin
 			}
 		}
 		// Legacy exact match
-		if opts.Predicate.Limit > 0 && len(opts.ResourceVersion) > 0 && opts.ResourceVersion != "0" {
+		if opts.Predicate.Limit > 0 && len(opts.ResourceVersion) > 0 && opts.ResourceVersion != "0" && utilfeature.DefaultFeatureGate.Enabled(features.ListFromCacheSnapshot) {
```

Comment:
However I don't agree with this fix. Semantic of this code is correct. The problem is in delegator.

The delegator code that handles consistent read can incorrectly change the semantic to legacy exact when limit is set.
https://github.com/kubernetes/kubernetes/blob/230b9f470a48acca6642f1c79428af878c6f8d4f/staging/src/k8s.io/apiserver/pkg/storage/cacher/delegator.go#L210-L217


My suggestion would be to add a line there to force `listLatestRV`. This can be done by adding:
```
opts.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
```

Reply:
Think I found another edge case that doesn't work. Consistent read that is a continue with negative rv.